### PR TITLE
Improvement in shape function computation

### PR DESCRIPTION
**Pull request by wenqing**
In the previous versions, the computations of shape functions are conducted within the local assembly over each element, which is redundant for the elements with the same type. 

In this PR, a class is introduced for computing shape functions _N_ and their derivatives with respect to the local coordinates d _N_  /d xi. The class is used to compute  _N_  and d _N_  /d xi at each integration points of all involved element types in a FEM computation, and it also holds all the computation results for later usage in the process of Neumann BC and the local assembly. During the local assembly, the derivatives of shape functions with respect to the global local coordinates are computed and cached before any numerical integration happens. By this way, 
- the redundancy in the shape function computation is reduced significantly, 
- the associated memory usage is less.

For the problem with big mesh or many time steps, the speedup of this improvement vs the previous versions could be distinct (20% has been reported).

Since all local shape function computations have been replaced with the cached data  acquisition, all benchmarks give tiny errors in their results against to that obtained by the previous versions.

### CODE REVIEW

File: FEM/ShapeFunctionPool.h [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+class ShapeFunctionPool
+{
+public:
+	/*!
+		\param elem_types          All involved element types.
+		\param quadrature          Numerical integration object.
+		\param num_sample_gs_pnts  Number of sample Gauss points.
+	*/
+	ShapeFunctionPool(const std::vector<MshElemType::type>& elem_types,
+		              CElement& quadrature, const int num_sample_gs_pnts);
```
Comment (norihiro-w) [inline FEM/ShapeFunctionPool.h]: Why is `CElement` object named  `quadrature`?

File: FEM/ShapeFunctionPool.h [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	/// Get the values of the gradient of shape function of an element type
+	double* getGradShapeFunctionValues(const MshElemType::type elem_type) const;
+
+	/// Get the gradient of shape function values at the element centroid of an element type
+	double* getGradShapeFunctionCenterValues(const MshElemType::type elem_type) const;
+
+private:
+	/// Results of shape functions of all integration points.
+	std::vector<double*> _shape_function;
```
Comment (norihiro-w) [inline FEM/ShapeFunctionPool.h]: you can use `vector<vector<double> >`. with this, you don't need to care about memory leak.

File: FEM/ShapeFunctionPool.h [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	/// Results of shape functions of all integration points at element centroid.
+	std::vector<double*> _shape_function_center; 
+	/// Results of the gradient of shape functions with respect to
+	/// local coordinates of all integration points.
+	std::vector<double*> _grad_shape_function; 
+	/// Results of the gradient of shape functions of all integration points at element centroid.
+	std::vector<double*> _grad_shape_function_center; 
+
+	void computeQuadratures(const std::vector<MshElemType::type>& elem_types,
+		                    const int num_elem_nodes[2][MshElemType::LAST],
```
Comment (norihiro-w) [inline FEM/ShapeFunctionPool.h]: indentation

Comment (norihiro-w) [inline FEM/ShapeFunctionPool.h]: what does `2` mean?

File: MSH/MSHEnums.h [hunk 4]
```diff
@@ -30,7 +30,9 @@ struct MshElemType
 		TETRAHEDRON = 5,
 		PRISM = 6,
 		PYRAMID = 7,
-		INVALID = -1
+		QUAD8 = 8,
+		INVALID = -1,
+		LAST = QUAD8
```
Comment (norihiro-w) [inline MSH/MSHEnums.h]: can you rename LAST to more meaningful name, e.g. N_ELEMENT_TYPES or MAX_ELEMENT_TYPE_ID?

File: FEM/ShapeFunctionPool.cpp [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	for (std::size_t i=0; i<elem_types.size(); i++)
+	{
+		const MshElemType::type e_type = elem_types[i];
+		if (e_type == MshElemType::INVALID)
+			continue;
+
+		const int type_id = static_cast<int>(e_type) - 1;
+		quadrature.ConfigShapefunction(e_type);
+
+    	const int nnodes = num_elem_nodes[order-1][type_id];
```
Comment (norihiro-w) [inline FEM/ShapeFunctionPool.cpp]: indentation

Comment (wenqing) [inline FEM/ShapeFunctionPool.cpp]: It has already been adjusted in the other commits.

File: FEM/ShapeFunctionPool.h (hunk 1 above)
Comment (wenqing) [inline FEM/ShapeFunctionPool.h]: The name of CElement is not proper, the class actually provides the computation of shape functions and their derivatives etc. Therefore I the second argument is named as ` quadrature`.

File: FEM/ShapeFunctionPool.h (hunk 3 above)
Comment (wenqing) [inline FEM/ShapeFunctionPool.h]: The indentation has been adjusted in the other commits of this PR.

`num_elem_nodes[2][MshElemType::LAST]` is a 2D array with size of 2 and MshElemType::LAST. Therefore `2` is the size of row of that array.

File: FEM/fem_ele.cpp [hunk 6]
```diff
@@ -112,22 +114,22 @@ CElement::CElement(int CoordFlag, const int order)
...
-	delete[] shapefct;
-	delete[] shapefctHQ;
-	delete[] dshapefct;
-	delete[] dshapefctHQ;
-	Jacobian = NULL;
-	shapefct = NULL;
-	dshapefct = NULL;
-	dshapefctHQ = NULL;
-	shapefctHQ = NULL;
+	if (Jacobian)
```
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: no need to check

File: FEM/fem_ele.cpp [hunk 7]
```diff
@@ -59,39 +58,42 @@ CElement::CElement(int CoordFlag, const int order)
...
-			dshapefct = new double[24];
-			dshapefctHQ = new double[60];
+			dim_jacobian = 9;
+			size_dshapefct = 24 * max_intgration_points;
+			size_dshapefctHQ = 60 * max_intgration_points;
 			//
 			break;
 	}
+
+	Jacobian = new double[dim_jacobian];
```
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: can add prefix `_` as this is a member variable?

File: FEM/fem_ele.cpp [hunk 8]
```diff
@@ -172,6 +175,7 @@ void CElement::ConfigElement(CElem* MElement, const int nquadrature_points, bool
 	{
 		if (dim != ele_dim)
 		{
+			dim_grad = dim;
```
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: isn't it possible to move this to line 164?

File: FEM/fem_ele.cpp [hunk 9]
```diff
@@ -503,47 +555,52 @@ double CElement::elemnt_average(const int idx, CRFProcess* m_pcs, const int orde
    01/2006 WW Axisymmtry
    09/2006 WW 1D element in 3D
 **************************************************************************/
-double CElement::computeJacobian(const int order)
+void CElement::computeJacobian(const int gp, const int order, const bool inverse)
```
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: can you rename `inverse` to `need_inverse`?

File: FEM/fem_ele.cpp [hunk 10]
```diff
@@ -1236,9 +1422,10 @@ double CElement::CalcXi_p()
 		}
 		r = 1.0 / Xi_p;
 		Xi_p = r;
+		return Xi_p;
 	}
-
-	return Xi_p;
+	else
+		return 0.;
```
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: indentation

File: FEM/fem_ele.h [hunk 11]
```diff
@@ -170,18 +217,34 @@ class CElement
 	int Order;
 	size_t ele_dim; // Dimension of element
 	size_t dim; // Dimension of real dimension
+	size_t dim_grad; // Gradient dimension
```
Comment (norihiro-w) [inline FEM/fem_ele.h]: i am not sure what is "gradient dimension". Is it different from the spacial dimension?

File: FEM/problem.cpp [hunk 12]
```diff
@@ -614,9 +618,25 @@ Problem::~Problem()
 	delete m_vec_BRNS;
 #endif
 
-#if defined(USE_PETSC) || defined(USE_MPI) || defined(USE_MPI_PARPROC) || defined(USE_MPI_REGSOIL) \
-    || defined(USE_MPI_GEMS)
-	if (mrank == 0)
+	if (_line_shapefunction_pool)
+	{
+		if (_line_shapefunction_pool == _quadr_shapefunction_pool)
```
Comment (norihiro-w) [inline FEM/problem.cpp]: you can write here as

```
		if (_line_shapefunction_pool == _quadr_shapefunction_pool)
			_quadr_shapefunction_pool = NULL;
		delete _line_shapefunction_pool;
		_line_shapefunction_pool = NULL;
```

File: FEM/problem.h [hunk 13]
```diff
@@ -103,8 +114,13 @@ class Problem
 
 	// Print flag
 	bool print_result;
-	// Processes
 
+	/// Caches for shape functions and their derivatives with respect to
+	/// the local coordinates.
+	FiniteElement::ShapeFunctionPool* _line_shapefunction_pool;
```
Comment (norihiro-w) [inline FEM/problem.h]: line -> linear

File: FEM/rf_num_new.cpp [hunk 14]
```diff
@@ -116,7 +116,7 @@ CNumerics::CNumerics(string name)
 		cpl_error_tolerance[i] = -1.0; // JT2012: should not default this. Should always be entered by user!
 	//
 	// ELE
-	ele_gauss_points = 3;
+	ele_gauss_points = 2;
```
Comment (norihiro-w) [inline FEM/rf_num_new.cpp]: why did you change the default number of gauss points? This should influence lots of benchmarks.

File: FEM/rf_pcs.cpp [hunk 15]
```diff
@@ -1067,6 +996,84 @@ void initializeConstrainedProcesses(std::vector<CRFProcess*>& pcs_vector)
 	}
 }
 
+void CRFProcess::SetBoundaryConditionAndSourceTerm()
+{
+	std::string pcs_type_name(convertProcessTypeToString(this->getProcessType()));
+
+	if (pcs_type_name_vector.size() && pcs_type_name_vector[0].find("DYNAMIC") != string::npos) // WW
```
Comment (norihiro-w) [inline FEM/rf_pcs.cpp]: please use `!pcs_type_name_vector.empty()` for the first condition

File: FEM/rf_pcs.cpp [hunk 16]
```diff
@@ -10659,8 +10665,7 @@ void CRFProcess::CalcSecondaryVariablesUnsaturatedFlow(bool initial)
 			if (elem->GetMark()) // Marked for use
 			{
 				elem->SetOrder(false);
-				fem->ConfigElement(elem, m_num->ele_gauss_points, false);
-				fem->CalcSatution();
+				fem->CalcSatuation(*elem);
```
Comment (norihiro-w) [inline FEM/rf_pcs.cpp]: typo -> `CalcSaturation`

File: FEM/rf_random_walk.cpp [hunk 17]
```diff
@@ -794,8 +794,8 @@ void RandomWalk::InterpolateVelocityOfTheParticleByInverseDistance(Particle* A)
 					P[p] = crossroad->plane[k].norm[p];
```
Comment (norihiro-w) [inline FEM/rf_random_walk.cpp]: actually there is no real change in rf_random_walk.h and .cpp. Can you revert the changes? The code formatting changes should be done in another PR.

File: MSH/MSHEnums.h (hunk 4 above)
Comment (wenqing) [inline MSH/MSHEnums.h]: Renamed.

File: FEM/ShapeFunctionPool.h (hunk 2 above)
Comment (wenqing) [inline FEM/ShapeFunctionPool.h]: replaced.

File: FEM/fem_ele.cpp (hunk 6 above)
Comment (wenqing) [inline FEM/fem_ele.cpp]: Removed.

File: FEM/fem_ele.cpp (hunk 7 above)
Comment (wenqing) [inline FEM/fem_ele.cpp]: added.

File: FEM/fem_ele.cpp (hunk 8 above)
Comment (wenqing) [inline FEM/fem_ele.cpp]: No, it can't be moved. It is for inclined elements. The dimensions of local and the global coordinates are different. For example, for a regular line element,  dim_grad = ele_dim = 1, while for a inclined line element in 3 D space,   dim_grad = 3.

File: FEM/fem_ele.cpp (hunk 9 above)
Comment (wenqing) [inline FEM/fem_ele.cpp]: Changed.

File: FEM/fem_ele.cpp (hunk 10 above)
Comment (wenqing) [inline FEM/fem_ele.cpp]: It is a tab space.

File: FEM/fem_ele.h (hunk 11 above)
Comment (wenqing) [inline FEM/fem_ele.h]: The dimensions of local and the global coordinates are different. For example, for a regular line element, dim_grad = ele_dim = 1, while for a inclined line element in 3 D space, dim_grad = 3.

File: FEM/problem.cpp (hunk 12 above)
Comment (wenqing) [inline FEM/problem.cpp]: Changed.

File: FEM/problem.h (hunk 13 above)
Comment (wenqing) [inline FEM/problem.h]: Changed.

File: FEM/rf_num_new.cpp (hunk 14 above)
Comment (wenqing) [inline FEM/rf_num_new.cpp]: In the previous versions,  ele_gauss_points is fixed to 2 for line element and 3 for quad and hex.
Just changed it to 3. Now all benchmarks with line element get longer computation time and better accuracy in solutions than before.

File: FEM/rf_pcs.cpp (hunk 15 above)
Comment (wenqing) [inline FEM/rf_pcs.cpp]: Replaced.

File: FEM/rf_pcs.cpp (hunk 16 above)
Comment (wenqing) [inline FEM/rf_pcs.cpp]: Corrected.

File: FEM/rf_random_walk.cpp (hunk 17 above)
Comment (wenqing) [inline FEM/rf_random_walk.cpp]: Discarded the format changes.

File: FEM/rf_react.cpp [hunk 18]
```diff
@@ -1766,16 +1767,16 @@ void CRFProcess::InterpolateTempGP(CRFProcess* m_pcs, std::string name)
 // MX
 void CRFProcess::ExtropolateTempGP(CRFProcess* m_pcs, std::string name)
 {
-	MshElemType::type EleType;
+	//	MshElemType::type EleType;
```
Comment (wenqing) [inline FEM/rf_react.cpp]: removed.

File: FEM/ShapeFunctionPool.h (hunk 2 above)
Comment (wenqing) [inline FEM/ShapeFunctionPool.h]: After replacement. Just be curious about performance with the replacement, and conducted a test with hm_foot_tet . The computation time increases 6 s (868 s vs 874 s), and this increment is tolerable.

File: FEM/fem_ele.h (hunk 11 above)
Comment (norihiro-w) [inline FEM/fem_ele.h]: thanks. can you rename those variables, e.g. ele_local_dim and ele_global_dim? I don't get what "grad" means.

File: FEM/fem_ele.cpp (hunk 8 above)
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: i see

File: FEM/fem_ele.cpp (hunk 10 above)
Comment (norihiro-w) [inline FEM/fem_ele.cpp]: sorry I couldn't see it

Comment (wenqing) [inline FEM/fem_ele.cpp]: If compare line 1427 and 1428, you will see that the indentation is OK. Line 1427 has an indentation of one tab, while line 1428 has two tabs of the indentation.  It looks like that one tab covers 8 charter space on this page.

I will apply clang-format once this PR is merged.

File: FEM/fem_ele.h (hunk 11 above)
Comment (wenqing) [inline FEM/fem_ele.h]: Just changed the name in commit ec6186e

File: FEM/CMakeLists.txt [hunk 19]
```diff
@@ -102,6 +102,8 @@ set( SOURCES
 	Stiff_Bulirsch-Stoer.cpp
 	tools.cpp
 	vtk.cpp
+	ShapeFunctionPool.h
```
Comment (norihiro-w) [inline FEM/CMakeLists.txt]: move this to HEADERS

**Comment by norihiro-w** [comment]
@wenqing This PR looks ok besides one minor comment. Have you checked why Jenkins is failing?

File: FEM/CMakeLists.txt (hunk 19 above)
Comment (wenqing) [inline FEM/CMakeLists.txt]: Changed.

**Comment by wenqing** [comment]
@norihiro-w Jenkin falied in the result comparison. Almost all results of the benchmarks have tiny difference against their references. I have checked these results, which look good. I wonder whether we should ask the people in charge of the benchmarks for a quick verification.

**Comment by norihiro-w** [comment]
Jenkins, test this please!

**Comment by norihiro-w** [comment]
@wenqing Jenkins is not working now. I think Lars and WKDV are working on it.

**Comment by bilke** [comment]
Still no response from WKDV..

**Comment by bilke** [comment]
Jenkins is working again!

**Comment by norihiro-w** [comment]
@bilke thanks!

**Comment by wenqing** [comment]
@norihiro-w  Several benchmarks that failed since the source code removed a ode solver (see issue #37). Jenkins needs this PR to be updated (ufz/ogs5-benchmarks_ref#2). Can we have a solution like that in ogs6/data such that Jenkins uses a branch of ufz/ogs5-benchmarks_ref to run the benchmarks? This can avoid to merge the changed ufz/ogs5-benchmarks_ref before merging the PR of the source code.

**Comment by bilke** [comment]
Sorry, Jenkins is down again. Ben will be back on Tuesday so we have to wait until then..

**Comment by norihiro-w** [comment]
@wenqing the new Jenkins results are available 
https://svn.ufz.de:8443/job/OGS-5/job/ufz/branch/PR-38/15/

**Comment by wenqing** [comment]
@norihiro-w 
Forgot to update. According to Lars' suggestion, the following modules were load on envinf (same as those defined in script/jekins/linux.grovy) for the test in my account
```
module load cmake/3.1.3-1
module load gcc/4.8.1-3
module load boost/1.55.0-4
```
The failed benchmarks are the same as that by the master of the repository:
```
The following tests FAILED:
	 19 - BG-ECLIPSE_DUMUX/1phase_radialflow_1phase_transport/1pf_1pt (Failed)
	 20 - BG-ECLIPSE_DUMUX/2phase_flow_2phase_tracertransport/2pf_2pt (Failed)
	 22 - BG-ECLIPSE_DUMUX/kinetic_CO2phase_generation_E100/CO2phase_gen_E100 (Failed)
	 23 - BG-ECLIPSE_DUMUX/kinetic_CO2phase_generation_E300/CO2phase_gen_E300 (Failed)
	 24 - BG-C/2d_Cl_transport_Clay/Nuklidtransport (Failed)
	 28 - CB-C/Engesgaard/Kin/slow_kin_pqc/pds (Failed)
	 29 - CB-C/Engesgaard/Kin/slow_kin_pqc_krc/pds (Failed)
	 31 - CB-C/NAPL-dissolution/1D_NAPL-diss_flow/lin/1D_TPF_resS_flow (Failed)
	 78 - MW-DENSITY-DEPENDENT_FLOW/goswami-clement/constrBC_PressAsHead_tri/HM (Failed)
	 79 - MW-DENSITY-DEPENDENT_FLOW/goswami-clement/wholeBC_PressAsHead_tri/HM (Failed)
Errors while running CTest
```

**Comment by wenqing** [comment]
test this please

**Comment by norihiro-w** [comment]
@wenqing can you rebase this? I'll merge it. then let's see which benchmarks are still failing

**Comment by wenqing** [comment]
@norihiro-w  Rebased.
